java/ql/src/utils/model-generator/CaptureSummaryModels.ql

```diff
@@ -0,0 +1,233 @@
+    not (fa.getField().isStatic() and fa.getField().isFinal()) and
+    returnNode.getEnclosingCallable() = api and
+    fa.getCompilationUnit() = api.getCompilationUnit() and
+    isRelevantType(api.getReturnType()) and
+    not api.getDeclaringType() instanceof EnumType and
+    TaintTracking::localTaint(DataFlow::exprNode(fa), returnNode)
+  |
+    result = asTaintModel(api, "Argument[-1]", asOutput(api, returnNode))
+  )
+}
+
+string asOutput(Callable api, ReturnNodeExt node) {
+  if node.getKind() instanceof ValueReturnKind
+  then result = "ReturnValue"
+  else
+    result = parameterAccess(api.getParameter(node.getKind().(ParamUpdateReturnKind).getPosition()))
+}
+
+class FieldAssignment extends AssignExpr {
+  FieldAssignment() { exists(Field f | f.getAnAccess() = this.getDest()) }
+}
+
+class ParameterToFieldConfig extends TaintTracking::Configuration {
+  ParameterToFieldConfig() { this = "ParameterToFieldConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof DataFlow::ParameterNode and
+    isRelevantType(source.getType())
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(FieldAssignment a |
+      a.getSource() = sink.asExpr() and
+      a.getDest().(VarAccess).getVariable().getCompilationUnit() =
+        sink.getEnclosingCallable().getCompilationUnit()
+    )
+  }
+}
+
+/**
+ * Captures APIs that accept input and store them in a field.
+ * Example:
+ * ```
+ * public class Foo {
+ *  private String tainted;
+ *  public void doSomething(String input) {
+ *    tainted = input;
+ *  }
+ * ```
+ * Captured Model:
+ * `p;Foo;true;doSomething;(String);Argument[0];Argument[-1];taint`
+ */
+string captureFieldFlowIn(Callable api) {
+  exists(DataFlow::ParameterNode source, DataFlow::ExprNode sink, ParameterToFieldConfig config |
+    sink.asExpr().getEnclosingCallable().getDeclaringType() =
+      source.asParameter().getCallable().getDeclaringType() and
+    not api.isStatic() and
+    config.hasFlow(source, sink) and
+    source.asParameter().getCallable() = api
+  |
+    result =
+      asTaintModel(api, "Argument[" + source.asParameter().getPosition() + "]", "Argument[-1]")
+  )
+}
+
+class ParameterToReturnValueTaintConfig extends TaintTracking::Configuration {
+  ParameterToReturnValueTaintConfig() { this = "ParameterToReturnValueTaintConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    exists(Callable api |
+      source instanceof DataFlow::ParameterNode and
+      api = source.asParameter().getCallable() and
+      isRelevantType(api.getReturnType()) and
+      isRelevantType(source.asParameter().getType())
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof ReturnNodeExt }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    node2.asExpr().(ConstructorCall).getAnArgument() = node1.asExpr() and
+    node1.asExpr().(Argument).getCall().getCallee().fromSource()
```

Comment:
What's the purpose of this?

Reply:
These are required to track the flow from the input parameters into returned objects.
These are the relevant models
```
| p;Factory;false;create;(String);;Argument[0];ReturnValue;taint |
| p;Factory;false;create;(String,int);;Argument[0];ReturnValue;taint |
| p;PrivateFlowViaPublicInterface;true;createAnSPI;(File);;Argument[0];ReturnValue;taint |
```